File: python/robotClient.py
import socket
import json
import threading
import Define
from representedData import RepresentedData
from signature import Signature
import time

class RobotClient ():
    SERVERS = [] # contains a tuple of (ip, port)
    QUORUM = 2

    ID = -1 # id do cliente. serve para identificar qual chave usar no RSA


    LOCK = threading.Lock()
    REQUEST_CODE = 0
    RESPONSES = [] # armazena tuplas (value, timestamp, data_signature, client_id, server)
    OUT_DATED_SERVERS = []


    SEMAPHORE = threading.Semaphore(0)

    VERBOSE = 0
    CERT_PATH = ''
    RESULTS_PATH = ''

    exit = False

    NUMBER_OF_EXECUTIONS = 0
    OPERATION_TIMERS = [] #array with timers of operations
    INIT_TIME = 0
    FINAL_TIME = 0

    '''
    Client constructor.
    param: id - Client id
    param: servers - Array with servers(ip+port)
    param: verbose - Verbose level: 0 - no print, 1 - print important, 2 - print all  
    param: cert_path - Path to certificates
    param: repeat_operations - number of times to repeat operation
    param: operation - read or write
    param: results_path - Path to save result txt
    '''
    def __init__(self, id, servers, verbose, cert_path, repeat_operations, operation, results_path):
        self.ID = id
        self.SERVERS = servers
        self.VERBOSE = verbose
        self.CERT_PATH = cert_path
        self.RESULTS_PATH = results_path

        print('Client ' + Define.plataform + " " + str(self.ID) + ' running...')
        #self.initUserInterface()
        self.makeRequests(repeat_operations, operation)


    '''
    Makes automated requests to servers
    param: (int) n - Number of times to send requests
    param: (string) type - read or write operarion
    '''
    def makeRequests(self, n, type):
        self.NUMBER_OF_EXECUTIONS = n
        self.INIT_TIME = time.time()

        i = 0
        if type == 'read':
            while (i < n):
                init_time = time.time()
                self.read()
                final_time = time.time()
                self.OPERATION_TIMERS.append(final_time-init_time)
                i += 1

            self.FINAL_TIME = time.time()
            self.writeExecutionInfo(self.RESULTS_PATH + 'client_' + str(self.ID) + '_read.txt')

        elif type == 'write':
            while (i < n):
                init_time = time.time()
                data = RepresentedData.getFakeData(200)  # 100kb
                self.write(data)
                final_time = time.time()
                self.OPERATION_TIMERS.append(final_time-init_time)
                i += 1

            self.FINAL_TIME = time.time()
            self.writeExecutionInfo(self.RESULTS_PATH + 'client_' + str(self.ID) + '_write.txt')


    '''
    Writes basic infos of execution in file
    param: (string) path - File's path to write info
    '''
    def writeExecutionInfo(self, path):
        file = open(path, 'w')

        file.write(Define.execution_file_header)
        infos = str(self.NUMBER_OF_EXECUTIONS) + ';' + str(self.getAverageOperationTime()) + ';' \
                + str(self.INIT_TIME) + ';' + str(self.FINAL_TIME)
        file.write(infos)
        file.close()

    '''
    Calculates average time for operations
    return: (float) average_time - Avarage time for operations
    '''
    def getAverageOperationTime(self):
        count = 0
        for time in self.OPERATION_TIMERS:
            count = count + time

        return count/len(self.OPERATION_TIMERS)

    '''
    Shows user inteface.
    '''
    def initUserInterface(self):
        while not self.exit:
            choice = 0
            while choice != '1' and choice != 1 and choice != '2' and choice != 2 and choice != '3' and choice != 3:
                self.cleanScreen()
                print ('*********************************\n')
                print ('O que deseja fazer?\n')
                print ('1 - Escrever valor na variavel\n')
                print ('2 - Ler valor da variavel\n')
                print ('3 - Sair\n')
                print ('*********************************\n')
                choice = input()

            if choice == '1' or choice == 1:
                self.cleanScreen()

                data = RepresentedData.getData()

                if data != '':
                    init_time = time.time()
                    self.write(data)
                    final_time = time.time()
                    print('Write function time: %.2f seconds' % (final_time - init_time))

            elif choice == '2' or choice  == 2:
                init_time = time.time()
                self.read()
                final_time = time.time()
                print ('Read function time: %.2f seconds' % (final_time-init_time))

            elif choice == '3' or choice == 3:
                self.terminate()

        print('Adeus')


    '''
    Clean screen.
    '''
    def cleanScreen(self):
        pass
        #todo os.system('cls' if os.name == 'nt' else 'clear')


    '''
    Writes a value on servers.
    param: value - Value to be written in servers (dictionary from RepresentedData class)
    '''
    def write(self, value):
        timestamp = self.readTimestamp()
        timestamp = self.incrementTimestamp(timestamp)

        data_signature = Signature.signData(Signature.getPrivateKey(-1,self.ID%2, self.CERT_PATH), value+str(timestamp))

        for server in self.SERVERS:
            threading.Thread(target=self.writeOnServer, args=(server, value, timestamp, data_signature, self.ID, self.REQUEST_CODE)).start()

        with self.LOCK:
            self.REQUEST_CODE = self.REQUEST_CODE + 1


    '''
    Read Timestamps from servers
    return: (int) Timestamp value; -1 if cant get timestamp.
    '''
    def readTimestamp(self):
        with self.LOCK:
            self.RESPONSES = []
            self.OUT_DATED_SERVERS = []

        if self.VERBOSE > 0:
            print("Lendo timestamp dos servidores....")
        for server in self.SERVERS:
            threading.Thread(target=self.readTimestampFromServer, args=(server, self.REQUEST_CODE)).start()

        wasReleased = self.SEMAPHORE.acquire(True, Define.timeout)

        if wasReleased:
            with self.LOCK:
                self.REQUEST_CODE = self.REQUEST_CODE + 1

            if (len(self.RESPONSES) >= self.QUORUM):
                timestamp = self.analyseTimestampResponse(self.RESPONSES)

                if self.VERBOSE > 0:
                    print('Li o dado do server:')
                    print("Timestamp: " + str(timestamp))

                return timestamp

            else:
                if self.VERBOSE > 0:
                    print("ERRO NAO ESPERADO!!!!!. Nao foi possivel ler nenhum dado. O semaforo liberou mas nao teve quorum.")
                return -1

        else:
            if self.VERBOSE > 0:
                print("Nao foi possivel ler nenhum dado. Timeout da conexao expirado")
            return -1


    '''
    Read Data from servers. Print on screen the response.
    '''
    def read(self):
        with self.LOCK:
            self.RESPONSES = []
            self.OUT_DATED_SERVERS = []

            if self.VERBOSE > 0:
                print("Lendo dados dos servidores....")
        for server in self.SERVERS:
            threading.Thread(target=self.readFromServer, args=(server, self.REQUEST_CODE)).start()

        wasReleased = self.SEMAPHORE.acquire(True, Define.timeout)

        if wasReleased:
            with self.LOCK:
                self.REQUEST_CODE = self.REQUEST_CODE + 1

            if (len(self.RESPONSES) >= self.QUORUM):
                value, timestamp, data_signature, client_id = self.analyseResponse(self.RESPONSES)

                self.writeBack(value, timestamp, data_signature, client_id)

                data = RepresentedData(value)
                if self.VERBOSE > 0:
                    print('Li o dado do server:')
                    data.showInfo()
                    print ("Timestamp: " + str(timestamp))

            else:
                if self.VERBOSE > 0:
                    print ("ERRO NAO ESPERADO!!!!!. Nao foi possivel ler nenhum dado. O semaforo liberou mas nao teve quorum.")

        else:
            if self.VERBOSE > 0:
                print ("Nao foi possivel ler nenhum dado. Timeout da conexao expirado")


    '''
    Writes a value readed on servers that are outdated.
    param: value - Value to be written in servers (dictionary from RepresentedData class)
    param: timestamp - Timestamp from value
    param: data_signature - Signature from value+timestamp
    param: client_id - ID from client that written the value
    '''
    def writeBack(self, value, timestamp, data_signature, client_id):
        for server in self.OUT_DATED_SERVERS:
            if timestamp != -1 and client_id != -1:
                threading.Thread(target=self.writeOnServer, args=(server, value, timestamp, data_signature, client_id, self.REQUEST_CODE)).start()

        with self.LOCK:
            self.REQUEST_CODE = self.REQUEST_CODE + 1


    '''
    Sends the data to be written in a specific server.
    param: server - Server to send the request
    param: value - Value to be written
    param: timestamp - Timestamp from value
    param: data_signature - Signature from value+timestamp
    param: client_id - Id from the client that created the value
    param: request_code - Request ID identifier
    '''
    def writeOnServer(self, server, value, timestamp, data_signature, client_id, request_code):
            request = dict(type=Define.write, timestamp=timestamp, variable=value, request_code = request_code, client_id = client_id, data_signature = data_signature)
            requestJSON = json.dumps(request)

            TCPSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCPSocket.connect(server)
            if self.VERBOSE == 2:
                print('-----REQUEST SAINDO:-----' + requestJSON)
            TCPSocket.send(requestJSON.encode('utf-8'))

            messageFromServer, server = TCPSocket.recvfrom(2048)
            messageFromServerJSON = messageFromServer.decode('utf-8')
            if self.VERBOSE == 2:
                print('-----REQUEST CHEGANDO:-----' + messageFromServerJSON)

            messageFromServerJSON = json.loads(messageFromServerJSON)
            if messageFromServerJSON[Define.status] == Define.success:
                if self.VERBOSE > 0:
                    print('Variable updated')
            else:
                if self.VERBOSE > 0:
                    print('Error updating')


    '''
    Gets the data from a server and append in the responses array if possible
    param: server - Server to send the request
    param: request_code - Request ID identifier
    '''
    def readFromServer(self, server, request_code):
        messageFromServer = self.getValueFromServer(server, request_code)

        if messageFromServer[Define.status] == Define.success and messageFromServer[Define.request_code] == self.REQUEST_CODE:
            data = messageFromServer[Define.data]
            serverTimestamp = data[Define.timestamp]
            serverVariable = data[Define.variable]
            dataSignature = data[Define.data_signature]
            clientID = data[Define.client_id]

            with self.LOCK:
                if len(self.RESPONSES) < self.QUORUM-1:
                    self.RESPONSES.append((serverVariable, serverTimestamp, dataSignature, clientID, server))

                elif len(self.RESPONSES) == self.QUORUM-1:
                    self.RESPONSES.append((serverVariable, serverTimestamp, dataSignature, clientID, server))
                    self.SEMAPHORE.release()

                else:
                    #do nothing
                    if self.VERBOSE > 0:
                        print("Quorum ja encheu. Jogando request fora...")

        elif messageFromServer[Define.status] == Define.error:
            if self.VERBOSE > 0:
                print("Ocorreu algum erro na request")

        elif messageFromServer[Define.request_code] != self.REQUEST_CODE:
            if self.VERBOSE > 0:
                print("Response atrasada.")


    '''
    Gets the timestamp from a server and append in the responses array if possible
    param: server - Server to send the request
    param: request_code - Request ID identifier
    '''
    def readTimestampFromServer(self, server, request_code):
        messageFromServer = self.getTimestampFromServer(server, request_code)

        if messageFromServer[Define.status] == Define.success and messageFromServer[Define.request_code] == self.REQUEST_CODE:
            data = messageFromServer[Define.data]
            serverTimestamp = data[Define.timestamp]

            with self.LOCK:
                if len(self.RESPONSES) < self.QUORUM-1:
                    self.RESPONSES.append(serverTimestamp)

                elif len(self.RESPONSES) == self.QUORUM-1:
                    self.RESPONSES.append(serverTimestamp)
                    self.SEMAPHORE.release()

                else:
                    #do nothing
                    if self.VERBOSE > 0:
                        print("Quorum ja encheu. Jogando request fora...")

        elif messageFromServer[Define.status] == Define.error:
            if self.VERBOSE > 0:
                print("Ocorreu algum erro na request")

        elif messageFromServer[Define.request_code] != self.REQUEST_CODE:
            if self.VERBOSE > 0:
                print("Response atrasada.")


    '''
    Makes the request for getting the value registered at a specific server
    param: server - Server to send the request
    param: request_code - Request ID identifier
    '''
    def getValueFromServer(self, server, request_code):
        request = dict(type=Define.read, request_code =request_code, client_id = self.ID)
        requestJSON = json.dumps(request)
        TCPSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        TCPSocket.connect(server)
        if self.VERBOSE == 2:
            print('-----REQUEST SAINDO:-----' + requestJSON)
        TCPSocket.send(requestJSON.encode('utf-8'))

        messageFromServer, server = TCPSocket.recvfrom(2048)
        messageFromServerJSON = messageFromServer.decode('utf-8')
        if self.VERBOSE == 2:
            print('-----REQUEST CHEGANDO:-----' + messageFromServerJSON)

        return json.loads(messageFromServerJSON)


    '''
    Makes the request for getting the timestamp registered at a specific server
    param: server - Server to send the request
    param: request_code - Request ID identifier
    '''
    def getTimestampFromServer(self, server, request_code):
        request = dict(type=Define.read_timestamp, request_code=request_code, client_id=self.ID)
        requestJSON = json.dumps(request)
        TCPSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        TCPSocket.connect(server)
        if self.VERBOSE == 2:
            print('-----REQUEST SAINDO:-----' + requestJSON)
        TCPSocket.send(requestJSON.encode('utf-8'))

        messageFromServer, server = TCPSocket.recvfrom(2048)
        messageFromServerJSON = messageFromServer.decode('utf-8')
        if self.VERBOSE == 2:
            print('-----REQUEST CHEGANDO:-----' + messageFromServerJSON)

        return json.loads(messageFromServerJSON)


    '''
    Makes the client exit.
    '''
    def terminate(self):
        self.exit = True


    '''
    Increments the timestamp
    param: serverTimestamp - Highest timestamp from a quorum of servers
    return: (int) Incremented timestamp
    '''
    def incrementTimestamp(self, serverTimestamp):
        return serverTimestamp+1


    '''
    Gets the data with the highest timestamp. Maps the outdated servers too.
    param: response - Servers' responses
    return: (tuple) Tuple with the actual data (value, timestamp, signature, writter_id)
    '''
    def analyseResponse(self, responses):
        value = ''
        timestamp = -1
        repeatTimes = 0
        auxServers = []
        data_signature = ''
        client_id = -1

        for (rValue, rTimestamp, data_sign, r_client_id, server) in responses:
            if rTimestamp == -1 or r_client_id == -1:
                #nao ha dado escrito no servidor
                self.OUT_DATED_SERVERS.append(server)

            else:
                if Signature.verifySign(Signature.getPublicKey(-1, r_client_id%2, self.CERT_PATH), data_sign, rValue + str(rTimestamp)):
                    if (rTimestamp == timestamp):
                        repeatTimes = repeatTimes + 1
                        auxServers.append(server)

                    elif (rTimestamp > timestamp):
                        timestamp = rTimestamp
                        repeatTimes = 1
                        value = rValue
                        data_signature = data_sign
                        client_id = r_client_id
                        self.transferObjects(auxServers, self.OUT_DATED_SERVERS)
                        auxServers.append(server)

                    else:
                        self.OUT_DATED_SERVERS.append(server)

                else:
                    # assinatura invalida
                    self.OUT_DATED_SERVERS.append(server)

        return (value, timestamp, data_signature, client_id)


    '''
    Gets the highest timestamp from an array containing various timestamps.
    param: response - Servers' responses
    return: (int) Highest timestamp founded
    '''
    def analyseTimestampResponse(self, responses):
        timestamp = -1
        repeatTimes = 0

        for rTimestamp in responses:
            if (rTimestamp == timestamp):
                repeatTimes = repeatTimes + 1

            elif (rTimestamp > timestamp):
                timestamp = rTimestamp
                repeatTimes = 1

            else:
                pass

        return timestamp


    '''
    Transfer Objects from one array to other
    param: oldArray - Array where objects will leave
    param: newArray - Array where objects will come
    '''
    def transferObjects(self, oldArray, newArray):
        for obj in oldArray:
            oldArray.remove(obj)
            newArray.append(obj)

File: java_honest_writers/src/main/java/com/caykeprudente/RobotClient.java
package com.caykeprudente;

import com.sun.tools.javac.util.Pair;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Scanner;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class RobotClient {
    List<Pair<String, Integer>> servers; // contains a tuple of (ip,port)
    int quorum = 2;

    int id = -1; // id do client. serve para identificar qual chave usar no RSA

    Lock lock = new ReentrantLock();

    private int request_code = 0;

    List<ResponseData> responses;
    List<Pair<String, Integer>> out_dated_servers;

    Semaphore semaphore = new Semaphore(0);

    public int verbose = 0;
    public String cert_path = "";
    public String results_path = "";

    boolean exit = false;

    public int number_of_executions = 0;
    public List<Double> operation_timers = new ArrayList<Double>();
    public long init_time = 0;
    public long final_time = 0;


    /*
    Client constructor.
    param: id - Client id
    param: servers - Array with servers(ip+port)
    param: verbose - Verbose level: 0 - no print, 1 - print important, 2 - print all
    param: cert_path - Path to certificates
    param: repeat_operations - number of times to repeat operation
    param: operation - read or write
    param: results_path - Path to save result txt
    */
    public RobotClient(int id, List<Pair<String, Integer>> servers, int verbose, String cert_path, int repeat_operations,
                       String operation, String results_path) {
        this.id = id;
        this.servers = servers;
        this.verbose = verbose;
        this.cert_path = cert_path;
        this.results_path = results_path;

        System.out.println("Client " + Define.plataform + " " + id + "running....");

        //initUserInterface();
        makeRequests(repeat_operations, operation);
    }

    /*
    Makes automated requests to servers
    param: (int) n - Number of times to send requests
    param: (string) type - read or write operarion
     */
    private void makeRequests(int n, String type) {
        this.number_of_executions = n;
        this.init_time = System.currentTimeMillis();

        int i = 0;
        if (type.equals("read")) {
            while (i < n) {
                long init_time = System.currentTimeMillis();
                this.read();
                long final_time = System.currentTimeMillis();
                this.operation_timers.add((final_time-init_time)/1000.0); //save seconds in array
                i++;
            }
            this.final_time = new Date().getTime();

            this.writeExecutionInfo(this.results_path + "client_" + this.id + "_read.txt");
        }
        else if (type.equals("write")) {
            while (i < n) {
                long init_time = System.currentTimeMillis();
                String data = RepresentedData.getFakeData(200);
                this.write(data);
                long final_time = System.currentTimeMillis();
                this.operation_timers.add((final_time-init_time)/1000.0); //save seconds in array
                i++;
            }
            this.final_time = System.currentTimeMillis();

            this.writeExecutionInfo(this.results_path + "client_" + this.id + "_write.txt");
        }
    }


    /*
    Writes basic infos of execution in file
    param: (string) path - File's path to write info
     */
    private void writeExecutionInfo(String path) {
        try {
            PrintWriter writer = new PrintWriter(path, "UTF-8");
            writer.println(Define.execution_file_header);
            writer.println(this.number_of_executions + "|" + getAverageOperationTime() + "|" + this.init_time/1000.0 + "|" + this.final_time/1000.0);
            writer.close();
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }

    /*
    Calculates average time for operations
    return: (float) average_time - Avarage time for operations
     */
    private Double getAverageOperationTime(){
        Double count = 0d;

        for (Double time : this.operation_timers)
            count += time;

        return count/this.operation_timers.size();
    }


    /*
    Shows user interface.
     */
    private void initUserInterface() {
        while (!exit) {
            Character choice = '0';
            while (choice != '1' && choice != '2' && choice != '3') {
                cleanScreen();

                System.out.println ("*********************************");
                System.out.println ("O que deseja fazer?");
                System.out.println ("1 - Escrever valor na variavel");
                System.out.println ("2 - Ler valor da variavel");
                System.out.println ("3 - Sair");
                System.out.println ("*********************************");

                Scanner scanner = new Scanner(System.in);
                choice = scanner.next().charAt(0);
            }

            if (choice == '1') {
                cleanScreen();

                String data = RepresentedData.getData();
                if (data != null)
                    write(data);
            }
            else if (choice == '2') {
                read();
            }
            else if (choice == '3') {
                terminate();
            }
        }

        System.out.println("Adeus");
    }


    /*
    Clean screen.
     */
    private void cleanScreen() {
        //todo apagar terminal
    }


    /*
    Writes a value on servers.
    param: value - Value to be written in servers (dictionary from RepresentedData class)
    */
    private void write(String value) {
        int timestamp = readTimestamp();
        timestamp = incrementTimestamp(timestamp);

        String dataSignature = MySignature.signData(MySignature.getPrivateKey(-1, id % 2, this.cert_path), value+timestamp);

        for (Pair<String, Integer> server : servers) {
            ResponseData data = new ResponseData(value, timestamp, dataSignature, id, request_code, server);
            ClientHandler handler = new ClientHandler(this, ClientHandler.Function.write, data);
            Thread thread = new Thread(handler);
            thread.start();
        }

        request_code++;
    }


    /*
    Read Timestamps from servers
    return: (int) Timestamp value; -1 if cant get timestamp.
    */
    private int readTimestamp() {
        lock.lock();
        responses = new ArrayList<ResponseData>();
        out_dated_servers = new ArrayList<Pair<String, Integer>>();
        lock.unlock();

        if (this.verbose > 0)
            System.out.println("Lendo timestamp dos servidores....");

        for (Pair<String, Integer> server : servers) {
            ResponseData data = new ResponseData(null, 0, null, id, request_code, server);
            ClientHandler handler = new ClientHandler(this, ClientHandler.Function.readTimestamp, data);
            Thread thread = new Thread(handler);
            thread.start();
        }

        try {
            boolean wasReleased = semaphore.tryAcquire(Define.timeout, TimeUnit.SECONDS);

            if (wasReleased) {
                lock.lock();
                request_code++;
                lock.unlock();

                if (responses.size() >= quorum) {
                    int timestamp = analyseTimestampResponse(responses);

                    if(this.verbose > 0)
                        System.out.println("Li o timestamp do server: " + timestamp);

                    return timestamp;
                }
                else {
                    if(this.verbose > 0)
                        System.out.println("ERRO NAO ESPERADO!!!!!. Nao foi possivel ler nenhum dado. O semaforo liberou mas nao teve quorum.");

                    return -1;
                }
            }
            else {
                if(this.verbose > 0)
                    System.out.println("Nao foi possivel ler nenhum dado. Timeout da conexao expirado");
                return -1;
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
            return -1;
        }
    }


    /*
    Read Data from servers. Print on screen the response.
     */
    private void read() {
        lock.lock();
        responses = new ArrayList<ResponseData>();
        out_dated_servers = new ArrayList<Pair<String, Integer>>();
        lock.unlock();

        if(this.verbose > 0)
            System.out.println("Lendo dados dos servidores....");

        for (Pair<String, Integer> server : servers) {
            ResponseData data = new ResponseData(null, 0, null, id, request_code, server);
            ClientHandler handler = new ClientHandler(this, ClientHandler.Function.read, data);
            Thread thread = new Thread(handler);
            thread.start();
        }

        try {
            boolean wasReleased = semaphore.tryAcquire(Define.timeout, TimeUnit.SECONDS);

            if (wasReleased) {
                lock.lock();
                request_code++;
                lock.unlock();

                if (responses.size() >= quorum) {
                    ResponseData mostRecentData = analyseResponse(responses);

                    writeBack(mostRecentData);

                    RepresentedData visibleData = new RepresentedData(mostRecentData.value);
                    if(this.verbose > 0)
                        System.out.println("Li o dado do server no timestamp " + mostRecentData.timestamp + ": ");
                    visibleData.showInfo();
                }
                else {
                    if(this.verbose > 0)
                        System.out.println("ERRO NAO ESPERADO!!!!!. Nao foi possivel ler nenhum dado. O semaforo liberou mas nao teve quorum.");
                }
            }
            else {
                if(this.verbose > 0)
                    System.out.println("Nao foi possivel ler nenhum dado. Timeout da conexao expirado")    ;
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }


    /*
    Writes a value readed on servers that are outdated.
    param: value - Value to be written in servers (dictionary from RepresentedData class)
    param: timestamp - Timestamp from value
    param: data_signature - Signature from value+timestamp
    param: client_id - ID from client that written the value
    */
    private void writeBack(ResponseData data) {
        for (Pair<String, Integer> server : out_dated_servers) {
            if (data.timestamp != -1 && data.client_id != -1) {
                data.server = server;
                data.request_code = request_code;

                ClientHandler handler = new ClientHandler(this, ClientHandler.Function.write, data);
                Thread thread = new Thread(handler);
                thread.start();
            }
        }
        request_code++;
    }


    /*
    Makes the client exit.
     */
    private void terminate(){
        exit = true;
    }


    /*
    Increments the timestamp
    param: serverTimestamp - Highest timestamp from a quorum of servers
    return: (int) Incremented timestamp
    */
    private int incrementTimestamp(int timestamp) {
        return timestamp+1;
    }


    /*
    Gets the data with the highest timestamp. Maps the outdated servers too.
    param: response - Servers' responses
    return: (tuple) Tuple with the actual data (value, timestamp, signature, writter_id)
    */
    private ResponseData analyseResponse(List<ResponseData> responses) {
        String value = "";
        int timestamp = -1;
        int repeatTimes = 0;
        ArrayList<Pair> auxServers = new ArrayList<Pair>();
        String data_signature = "";
        int client_id = -1;

        for (ResponseData response : responses) {
            if (response.timestamp == -1 || response.client_id == -1) {
                //nao ha dado escrito no servidor em questao
                out_dated_servers.add(response.server);
            }
            else {
                if (MySignature.verifySign(MySignature.getPublicKey(-1, response.client_id % 2, this.cert_path), response.data_signature, response.value+response.timestamp)) {
                    if (response.timestamp == timestamp) {
                        repeatTimes = repeatTimes + 1;
                        auxServers.add(response.server);
                    }
                    else if (response.timestamp > timestamp) {
                        timestamp = response.timestamp;
                        repeatTimes = 1;
                        value = response.value;
                        data_signature = response.data_signature;
                        client_id = response.client_id;
                        transferObject(auxServers, out_dated_servers);
                        auxServers.add(response.server);
                    }
                    else
                        out_dated_servers.add(response.server);
                }
                else {
                    out_dated_servers.add(response.server);
                }
            }
        }

        return new ResponseData(value, timestamp, data_signature, client_id, 0, null);
    }


    /*
    Gets the highest timestamp from an array containing various timestamps.
    param: responses - Servers' responses
    return: (int) Highest timestamp founded
    */
    private int analyseTimestampResponse(List<ResponseData> responses) {
        int timestamp = -1;
        int repeatTimes = 0;

        for (ResponseData response : responses) {
            if (response.timestamp == timestamp)
                repeatTimes++;
            else if (response.timestamp > timestamp) {
                timestamp = response.timestamp;
                repeatTimes = 1;
            }
            else
                break;
        }

        return timestamp;
    }


    /*
    Transfer Objects from one array to other
    param: oldArray - Array where objects will leave
    param: newArray - Array where objects will come
    */
    private void transferObject(List oldArray, List newArray) {
        for (Object object : oldArray) {
            oldArray.remove(object);
            newArray.add(object);
        }
    }
}
